File: cole/day28_missing-number/missing-number.c
// the xor solution
int missingNumber(int *nums, int numsSize)
{
    int res = numsSize;
    for (int i = 0; i < numsSize; i++)
    {
        res ^= i;
        res ^= nums[i];
    }
    return res;
}

// the sum solution
int missingNumber(int *nums, int numsSize)
{
    int sum = 0;
    for (int *it = nums, *end = nums + numsSize; it < end; ++it)
    {
        sum += *it;
    }
    return (numsSize * (numsSize + 1) / 2) - sum;
}

File: brady/missing_number.java
class Solution {
  
    // O(n) time, O(1) space
    public int missingNumber(int[] nums) {
        // using short because n <= 10000 < 2^15 - 1

        // XOR all the nums, and all the numbers from 0 to nums.length (inclusive)
        short allNums = 0;
        short actual = 0;
        for (short i = 0; i < nums.length; ++i) {
            allNums ^= i;
            actual ^= nums[i];
        }
        allNums ^= nums.length;

        return allNums ^ actual;
    }
}
